src/element/element_stream_writer.rs

```diff
@@ -0,0 +1,697 @@
+            containers: vec![ContainerContext {
+                annotations: Annotations::empty(),
+                field_name: None,
+                container: Container::TopLevel,
+            }],
+            annotations: Annotations::empty(),
+            field_name: None,
+        }
+    }
+
+    fn current_container(&self) -> &ContainerContext {
+        // `self.containers` is never empty; it always has at least the top level.
+        self.containers.last().unwrap()
+    }
+
+    fn current_container_mut(&mut self) -> &mut ContainerContext {
+        // `self.containers` is never empty; it always has at least the top level.
+        self.containers.last_mut().unwrap()
+    }
+
+    fn take_current_annotations(&mut self) -> Annotations {
+        std::mem::replace(&mut self.annotations, Annotations::empty())
+    }
+
+    fn push_container(&mut self, container: Container) -> IonResult<()> {
+        let annotations = self.take_current_annotations();
+        let field_name = self.field_name.take();
+        self.containers.push(ContainerContext {
+            annotations,
+            field_name,
+            container,
+        });
+        Ok(())
+    }
+
+    fn pop_container(&mut self) -> IonResult<ContainerContext> {
+        if self.containers.len() <= 1 {
+            return illegal_operation("cannot step out of the top level");
+        }
+        // `self.containers` is never empty; it always has at least the top level.
+        Ok(self.containers.pop().unwrap())
+    }
+
+    fn set_current_annotations(&mut self, annotations: Annotations) {
+        self.annotations = annotations;
+    }
+
+    fn write_scalar(&mut self, value: Value) -> IonResult<()> {
+        let annotations = self.take_current_annotations();
+        let element = value.with_annotations(annotations);
+        self.write_element(element)
+    }
+
+    fn write_element(&mut self, element: Element) -> IonResult<()> {
+        let field_name = self.field_name.take();
+        match &mut self.current_container_mut().container {
+            Container::TopLevel => self.output_mut().extend(std::iter::once(element)),
+            Container::SExpression(seq) => seq.push(element),
+            Container::List(seq) => seq.push(element),
+            Container::Struct(fields) => {
+                if let Some(field_name) = field_name {
+                    fields.push((field_name, element));
+                } else {
+                    return illegal_operation("Values inside a struct must have a field name.");
+                }
+            }
+        };
+        Ok(())
+    }
+
+    fn to_symbol<A>(&self, sym_ref: A) -> IonResult<Symbol>
+    where
+        A: AsRawSymbolTokenRef,
+    {
+        match sym_ref.as_raw_symbol_token_ref() {
+            RawSymbolTokenRef::SymbolId(_symbol_id) => {
+                // TODO what to do with symbol ids?
+                todo!()
+            }
+            RawSymbolTokenRef::Text(txt) => Ok(Symbol::owned(txt)),
+        }
```

Comment:
I'm unsure what cases could lead to a symbol id being used here, and thus unsure how to handle this case. Error?

Reply:
Try to resolve the SID using the symbol table. In this case, that's probably just the [system symbol table](https://amazon-ion.github.io/ion-docs/docs/symbols.html#system-symbols). If you instantiate a `SymbolTable`, it'll contain the system symbols by default.

If the symbol ID's out of bounds (ie `>=10`), that's an `illegal_operation` error. If it's `0`, you can construct a `Symbol::unknown_text()`.

Reply:
Done